demisto_sdk/__main__.py

```diff
@@ -1916,6 +1916,99 @@ def openapi_codegen(**kwargs):
+    required=False)
+@click.option(
+    '-f', '--fix-code', is_flag=True, help='Fix the python code using autopep8', default=False)
+@click.option(
+    '-v', '--verbose', is_flag=True, help='Be verbose with the log output', default=False)
+def ansible_codegen(**kwargs):
+    """Generates a Cortex XSOAR Ansible integration given a integration configuration file.
+    In the first run of the command, an integration configuration file is created, which needs be modified with \
+        the details of the Ansible which will be packaged as a integration.
+    Then, the command is run a second time with the integration configuration to generate the actual integration files.
+    """
+
+    from demisto_sdk.commands.ansible_codegen.ansible_codegen import \
+        AnsibleIntegration
+    from demisto_sdk.commands.common.hook_validations.docker import \
+        DockerImageValidator
+
+    if not kwargs.get('output_dir'):
+        output_dir = os.getcwd()
+    else:
+        output_dir = kwargs['output_dir']
+
+    # Check the directory exists and if not, try to create it
+    if not os.path.exists(output_dir):
+        try:
+            os.mkdir(output_dir)
+        except Exception as err:
+            print_error(f'Error creating directory {output_dir} - {err}')
+            sys.exit(1)
+    if not os.path.isdir(output_dir):
+        print_error(f'The directory provided "{output_dir}" is not a directory')
+        sys.exit(1)
+
+    base_name = kwargs.get('base_name')
+    verbose = kwargs.get('verbose')
+    fix_code = kwargs.get('fix_code')
+
+    container_image = kwargs.get('container_image')
+    if not container_image:  # Click sets the value to None even if not provided by user.
+        container_image = f"demisto/ansible-runner:{DockerImageValidator.get_docker_image_latest_tag_request('demisto/ansible-runner')}"
+
+    config_file_path = kwargs.get('config_file')
+
+    integration = AnsibleIntegration(base_name=base_name, verbose=verbose, container_image=container_image,
+                                     output_dir=output_dir, config_file_path=config_file_path, fix_code=fix_code)
+
+    if not config_file_path:
+        integration.save_empty_config(output_dir)
+        config_path = os.path.join(output_dir, f'{base_name}_config.yml')
+        print_success(f'Created empty configuration file {config_path}. ')
+        command_to_run = f'demisto-sdk ansible-codegen -cf "{config_path}" -ci "{container_image}" -n "{base_name}" -o "{output_dir}"'
+        command_to_run += ' -v' if verbose else ''
+        command_to_run += ' -f' if fix_code else ''
+
+        click.echo(f'Run the command again with the created configuration file (after populating it): {command_to_run}')
+        sys.exit(0)
+
+    click.echo('Loading AnsibleCodegen configuration file...')
+    integration.load_config()
+
+    click.echo('Validating AnsibleCodegen configuration file...')
+    validate, validate_message = integration.validate()
```

Comment:
Why the validate is before saving the integration package? dosn't it make more sense to validate after creating the integration? Also, the `save_yaml` is called twice (in validate and in save_package)

Reply:
Good observation. The code generation process can be resource intensive (mostly disk) when it comes to the docs parsing, especially if there are hundreds of referenced Ansible modules. For the best user experience I wanted to perform as much validation prior to this compute intensive phase as possible so the process is aborted as soon as it is known that it will fail.

`save_yaml` is called twice because the AnsibleCodegen configuration file can contain native XSOAR integration structure. The AnsibleCodegen config items `category` and `parameters` are used without alteration in the generated integration files. Thus the first save is to create a barebones integration with these config items for validation purposes, and the second `save_yaml` is called to save the actual fully generated package.